lib/common/procfs.c

```diff
@@ -200,3 +200,28 @@ pcmk__procfs_pid2path(pid_t pid, char path[], size_t path_size)
     return EOPNOTSUPP;
 #endif
 }
+
+/*!
+ * \internal
+ * \brief Check whether process ID information is available from procfs
+ *
+ * \return true if process ID information is available, otherwise false
+ */
+bool
+pcmk__procfs_has_pids(void)
```

Comment:
I guess I'd have to mock pcmk__procfs_pid2path() for that? Does mocking our own functions work the same way?

Alternatively I could mock getpid() and use the same mock readlink() as the other function, but then would both functions need to be tested in the same file, or is there a way to share a mocked function between two files?

Reply:
> At the moment, there is no way to share mocked functions. It seems like it would be easy to do (just add a file that gets built into libcrmcommon_test only), but I think our usage of a single mocked function will vary enough from test to test that sharing things will be more complicated than we want.
> 
> On the other hand, looking through existing tests, the only weirdness I see is around __wrap_getenv. But, having all the functions in one place would take away the flexibility of being able to test the mocked function in different ways. Maybe some tests will want a system function to fail in one way, and others will want it to fail in another.

We could use a will_return() variable to select the desired behavior.

> 
> From a linking perspective, I think it would alright though. We would always have all the mocked versions linked into every test executable, but I think the only ones that would ever get used would be ones with -Wl,--wrap= compiler flags.

I think the problem would be that if a function uses a will_return() variable, every caller *has* to set it. For something like calloc() that would be impractical. In fact unit testing anything that needs to mock calloc() will probably be impractical for anything but leaf functions.

Reply:
In fact we already have it: lib/common/mock.c with the weak version of every wrapper. I think we just need to put the real wrappers in there.

Reply:
Double check that we're right, and that the mocked versions only get called if there's a -Wl,--wrap= compiler flag to activate them.

Reply:
Bah. Putting the wrapper definitions (which need to be able to call \_\_real\_\*) in mock.c means mock.c needs to be linked with --wrap for everything.

I have an idea I'm trying, though.